### PR TITLE
Creating ContentPart alternates automatically

### LINKED ISSUE #1127: Troubles accessing the Html property of a MarkdownPart in Liquid Syntax
There is probably an easy solution, I just don't get it...
What I've tried sofar (within a for loop in Liquid Syntax) is shown below:
- These examples give at least the unconverted markdown: 
`<p>{{ feature.MarkdownPart.Markdown | markdownify }}</p>`
`<p>{{ feature.MarkdownPart.Markdown }}</p>`
`<p> {% display feature.MarkdownPart %} </p>`
- Whenever I access the Html Property (as I understood from the documentation) it returns nothing:
`<p>{{ feature.MarkdownPart.Html }}</p>`
`<p> {% display feature.MarkdownPart.Html %} </p>`
`<p>{{ feature.MarkdownPart.Markdown.Html }}</p>`
Any help or reference to a document would be nice...

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #1127

### CODE REVIEW

**Comment by sebastienros** [comment]
@jtkech here I am creating new part alternates automatically. I also improved BagPart to be able to set a custom display type. And we can do `{% display Model.Content.Services %}` were `Services` is a named part.

This should help with the issue #1127 as we can not just call this display, and be able to create a custom template for it.

We still need a way to render a content item directly, something like `{% display service | build_display: "Detail" %}. Can you do it?

**Comment by sebastienros** [comment]
@agriffard you might want to test this PR, I found some bugs while running it, many type changed. But it adds new shiny part alternates that you can use to custom things.

**Comment by jtkech** [comment]
@sebastienros 

Ok, but i need to re-put my mind into all of this to be able to review this PR. Then i will do a `build_display` filter. Can't right now but i will start on it tomorrow.

File: src/OrchardCore.Modules/OrchardCore.Contents/Shapes.cs [hunk 1]
```diff
@@ -73,7 +73,7 @@ public void Discover(ShapeTableBuilder builder)
 
                     if (!String.IsNullOrEmpty(alternate))
                     {
-                        ((IShape)displayShape).Metadata.Alternates.Add(displayShape);
+                        displayShape.Metadata.Alternates.Add(alternate);
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Contents/Shapes.cs]: Indeed ;)

File: src/OrchardCore.Modules/OrchardCore.Liquid/Views/LiquidPart.Summary.cshtml [hunk 2]
```diff
@@ -1,5 +1,4 @@
 @using OrchardCore.Mvc.Utilities;
 @model OrchardCore.Liquid.ViewModels.LiquidPartViewModel
 
-@Html.Raw(Model.Html.RemoveTags(true).Substring(0, 200))
-
+@Html.Raw(Model.Html.RemoveTags(true).Ellipsize(200))
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Liquid/Views/LiquidPart.Summary.cshtml]: Yes

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs [hunk 3]
```diff
@@ -107,16 +107,21 @@ internal class ShapeAccessor : DelegateAccessor
...
 
-            foreach (var item in (o as Shape).Items)
-            {
-                if (item is IShape && item.Metadata.Type == n)
+                foreach (var item in shape.Items)
                 {
-                    return item;
+                    // The prefix is the Name of the part. It also works for named parts,
+                    // so we can access Model.Content.MyNamedPart
+                    if (item is IShape itemShape && itemShape.Metadata.Prefix == n)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs]: With some shapes, e.g when `.Type = "Contents__Metadata"`, we have `.Prefix = ""`. With some shapes, we have `.Prefix = null`. So, maybe we should also check `.Type`.

I also see the following in some places where the dot character could be a problem to access properties.

    Prefix = htmlFieldPrefix + "." + Prefix;
    Prefix = context.HtmlFieldPrefix + "." + Prefix;
    Prefix = typePartDefinition.Name + "." + partFieldDefinition.Name;

Humm, maybe we would have the same problem with fields as with named part, but here the prefix contains a dot.

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentPartDisplayDriverTPart.cs [hunk 4]
```diff
@@ -14,6 +15,39 @@ namespace OrchardCore.ContentManagement.Display.ContentDisplay
     /// <typeparam name="TPart"></typeparam>
     public abstract class ContentPartDisplayDriver<TPart> : DisplayDriverBase, IContentPartDisplayDriver where TPart : ContentPart, new()
     {
+        private ContentTypePartDefinition _typePartDefinition;
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentPartDisplayDriverTPart.cs]: I'm not sure as i'm reading this but could we have the same instance of `ContentPartDisplayDriver` used for the same part but for 2 different types? And / or for 2 different named part?

**Comment by agriffard** [comment]
What kind of alternates can I test for example?

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs (hunk 3 above)
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs]: Then I will add a `Name` property, which will default to the type, then the name, and can be anything else too.

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentPartDisplayDriverTPart.cs (hunk 4 above)
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplay/ContentPartDisplayDriverTPart.cs]: Even if it was the case, the value would be set again on the second usage. As long as you return a Shape.

**Comment by jtkech** [comment]
@sebastienros just re-tried, great work.

- I tried the blog and agency setups.

- Now we can access `Model.Content.Features` which is a `Castle.Proxies.BagPartViewModelProxy`

- We still can't access the following but we know why.

      Model.Content.Features.Content
      Model.Content.Features.ContentItems

- But i could do the following so i think we already can solve #1127

      {% display Model.Content.Features %}

Let me some time to see what we can do now with alternates. I will try to create a template to override the `Features BagPart`.

**Comment by sebastienros** [comment]
I think it's fine that we can't access `BagPartViewModelProxy`, it's a shape, it should be displayed and have a custom template. All the other data is available in the content item.

**Comment by jtkech** [comment]
@sebastienros ok.

- Good, i could override `BagPart__LandingPage__Features`.

- But as said before, now we would need to do a `build_display` liquid tag.

But, before doing this liquid tag, is it normal that `BuildDisplayAsync()` needs to be called through the template view? This in place of calling it through a driver.

**Comment by sebastienros** [comment]
> But, before doing this liquid tag, is it normal that BuildDisplayAsync() needs to be called through the template view? This in place of calling it through a driver.

That's a good question, which I asked myself by the past and again yesterday before filing the issues ...
One argument to let the template decide is that the display type could be changed. If we generated the shapes from the driver then we couldn't. And worst case we would generate the shapes, then generate them again with the required display type.

**Comment by jtkech** [comment]
Ok, i understand but now for bagpart we can specify the `Display Type`. But maybe you mean that in some cases we still want to change it on the fly.

**Comment by jtkech** [comment]
Anyway it's good to create the `build_display` tag.

**Pull request merged**